### PR TITLE
Store user preferences

### LINKED ISSUE #439: Make UserPreferences persistable and updatable in the Repository and Database
## Description
Make UserPrefernces persistable and updatable in the Repository and Database.

## Dependencies
#418 #420
## Definition of done
- [ ] Make UserPrefernces persistable and updatable
- [ ] Use this in the UpdateUserPreferences effect
- [ ] Add tests

### PR DESCRIPTION
**Pull request by Odrakir**
## Info
Adds the necessary code to store user preferences in `UserDefaults`.

### Summary
This extends the `UserPreferences` struct to make it `Codable` and then adds another property wrapper to store `Codable` objects in `UserDefaults` 

### Technical considerations
I tried to avoid creating another property wrapper for this and instead using the `UserDefault` one, but I didn't manage to make it work so I could save both plain types and `Codable` ones. Any ideas here are welcome.

### Why is it done like this and not some other way?
Alternatively user preferences could be stored in the DB, but I think this makes more sense in `UserDefaults`. This is done by the `Repository`, which has that dependency, so we can still sync it with the server and it's transparent for the rest of the app.

### Relationships
Closes #439 

### Pull Request checklist
- [x] The code does not hinder the app's performance
- [x] The code does not introduce regressions
- [x] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests

### Pointers
One commit at a time

### Review template

- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones

- [ ] The code considers light and dark theme
- [ ] The code considers RTL and does not mention left/right
- [ ] The code considers i18n
  - [ ] There are no hardcoded strings in the UI
  - [ ] All strings are included in every language we support OR there's an issue requesting new translations
- [ ] There are no hard-coded numbers referring to dimensions or font sizes
- [ ] The code considers UI elements that are similar and extracts their similarities into styles to avoid duplication

## Merge permissions
Go!

### CODE REVIEW

File: Modules/OtherServices/OtherServices/MockUserDefaultService.swift [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import Foundation
 
 #if DEBUG
-public class MockUserDefaultService: UserDefaultService {
+public class MockUserDefaultService: UserDefaultsService {
```
Comment (dshr) [inline Modules/OtherServices/OtherServices/MockUserDefaultService.swift:4]: Shouldn't this then be `MockUserDefaultsService`?

File: Modules/Settings/Settings/Common/SettingsReducer.swift [hunk 2]
```diff
@@ -15,7 +15,7 @@ public func createSettingsReducer(
     calendarService: CalendarService,
     platformInfo: PlatformInfo,
     feedbackAPI: FeedbackAPI,
-    userDefaultsService: UserDefaultService
+    userDefaultsService: UserDefaultsService
```
Comment (dshr) [inline Modules/Settings/Settings/Common/SettingsReducer.swift:18]: Why are we injecting it here? I thought it all had to go through the Repository?

File: Modules/TogglTrack/TogglTrack/AnalyticsReducer.swift [hunk 3]
```diff
@@ -6,7 +6,7 @@ import Settings
 import OtherServices
 
 func createAnalyticsReducer(analyticsService: AnalyticsService,
-                            userDefaultService: UserDefaultService)
+                            userDefaultService: UserDefaultsService)
```
Comment (dshr) [inline Modules/TogglTrack/TogglTrack/AnalyticsReducer.swift:9]: Same thing here

File: Modules/TogglTrack/TogglTrack/App/AppEnvironment.swift [hunk 4]
```diff
@@ -18,7 +18,7 @@ public struct AppEnvironment {
     public let analytics: AnalyticsService
     public let calendarService: CalendarService
     public let notificationsService: NotificationsService
-    public let userDefaultService: UserDefaultService
+    public let userDefaultService: UserDefaultsService
```
Comment (dshr) [inline Modules/TogglTrack/TogglTrack/App/AppEnvironment.swift:21]:
```suggestion
public let userDefaultsService: UserDefaultsService
```

File: Modules/Models/Models/UserPreferences.swift [hunk 5]
```diff
@@ -89,8 +89,49 @@ public enum DateFormat: Equatable {
...
+    }
+
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        try container.encode(selectedWorkspaceId?.rawValue, forKey: .selectedWorkspaceId)
+        var smartAlertsStateTimeBefore: Double?
+        if case let .enabled(timeBefore) = smartAlertsState {
+            smartAlertsStateTimeBefore = timeBefore
+        }
+        try container.encode(smartAlertsStateTimeBefore, forKey: .smartAlertsState)
```
Comment (sofokli1) [inline Modules/Models/Models/UserPreferences.swift:127]: why not put this in the `if` and drop the var altogether?

File: Modules/Settings/Settings/Common/SettingsReducer.swift (hunk 2 above)
Comment (Odrakir) [reply Modules/Settings/Settings/Common/SettingsReducer.swift:18]: It doesn't necessarily have to always go through the repo, we could, however split it in interfaces to make that difference clearer. I'll take a note.

**Comment by sofokli1** [comment]
Taking this over to fix and merge

**Pull request merged**